### DISCUSSION TITLE
ch9/ch9-02

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#212] [General]
# ch9/ch9-02

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch9/ch9-02.html

**Comment by LiZhongChen0073** [comment]
这两章不错，嘿嘿 确实在实力工作中在类似的地方栽过，学到了

**Comment by Etherealss** [comment]
在[stackoverflow](https://stackoverflow.com/a/14671462/18656443)底下的评论里也有不少人反对“拒绝可重入锁”这种设计。虽然说可重入锁确实存在一些问题和弊端，但不可重入的mutex难道就不会导致死锁之类的问题吗？它也会影响开发者的代码设计，例如影响封装效果、抬高代码复杂度。
本章提到的“不变性的深层含义”其实可能是一种强加在Metex之上的含义，而Metex本来不必承担这种含义……又或者，完全可以提出另一种锁类型来承担这种深层含义。不管怎么说，如果**仅凭文中和链接中的理由**就从语言层面拒绝可重入锁、而不是让开发者去选择，那实在令人感觉局限、不灵活。

**Comment by mefly2012** [comment]
使用def mutex.unlock会导致函数结束了才释放锁，如果函数逻辑小还好，如果锁的范围太大了，影响效率。
